jni/src/commons.cpp

```diff
@@ -107,3 +114,140 @@ int knn_jni::commons::getIntegerMethodParameter(JNIEnv * env, knn_jni::JNIUtilIn
+        float32x4_t sv = vdupq_n_f32(src[i]);
+        float16x4_t hv = vcvt_f16_f32(sv);
+        __fp16 lane = vget_lane_f16(hv, 0);
+        dst[i] = *reinterpret_cast<const uint16_t*>(&lane);
+    }
+
+#elif defined(__x86_64__)
+  #if defined(__AVX512F__)
+    // x86 AVX-512 bulk 16-wide
+    for (; i + 16 <= count; i += 16) {
+        __m512 v = _mm512_loadu_ps(&src[i]);
+        __m256i h = _mm512_cvtps_ph(v, _MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC);
+        _mm256_storeu_si256(reinterpret_cast<__m256i*>(&dst[i]), h);
+    }
+  #elif defined(__AVX2__) && defined(__F16C__)
+    // x86 AVX2+F16C bulk 8-wide
+    for (; i + 8 <= count; i += 8) {
+        __m256 v = _mm256_loadu_ps(&src[i]);
+        __m128i h = _mm256_cvtps_ph(v, _MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC);
+        _mm_storeu_si128(reinterpret_cast<__m128i*>(&dst[i]), h);
+    }
+  #else
+    #error "x86_64 must support AVX512F or AVX2+F16C"
+  #endif
+    // tail via F16C scalar
+    for (; i < count; ++i) {
+        __m128 sv = _mm_set_ss(src[i]);
+        __m128i hv = _mm_cvtps_ph(sv, _MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC);
+        dst[i] = static_cast<uint16_t>(_mm_cvtsi128_si32(hv));
+    }
+#else
+    #error "Only aarch64 or x86_64 supported"
+#endif
+
+    env->ReleasePrimitiveArrayCritical(fp16Array, dst_bytes, 0);
+    env->ReleasePrimitiveArrayCritical(fp32Array, src_f32, JNI_ABORT);
+}
+
+
+// FP16 → FP32
+void knn_jni::commons::convertFP16ToFP32(knn_jni::JNIUtilInterface* jniUtil,
+                                         JNIEnv* env,
+                                         jbyteArray fp16Array,
+                                         jfloatArray fp32Array,
+                                         jint count,
+                                         jint offset) {
+    if (count <= 0) return;
+
+    jfloat* dst_f32 = reinterpret_cast<jfloat*>(env->GetPrimitiveArrayCritical(fp32Array, nullptr));
+    jbyte* src_bytes = reinterpret_cast<jbyte*>(env->GetPrimitiveArrayCritical(fp16Array, nullptr));
+    float* dst = reinterpret_cast<float*>(dst_f32);
+    const uint16_t* src = reinterpret_cast<const uint16_t*>(src_bytes + offset);
+
+    int i = 0;
+
+#if defined(__aarch64__)
+    // ARM NEON bulk 8-wide
+    for (; i + 8 <= count; i += 8) {
+        float16x4_t h0 = vld1_f16(reinterpret_cast<const __fp16*>(&src[i + 0]));
+        float16x4_t h1 = vld1_f16(reinterpret_cast<const __fp16*>(&src[i + 4]));
+        float32x4_t v0 = vcvt_f32_f16(h0);
+        float32x4_t v1 = vcvt_f32_f16(h1);
+        vst1q_f32(&dst[i + 0], v0);
+        vst1q_f32(&dst[i + 4], v1);
+    }
+    // tail via NEON scalar broadcast
+    for (; i < count; ++i) {
+        __fp16 half = *reinterpret_cast<const __fp16*>(&src[i]);
+        float32x4_t fv = vcvt_f32_f16(vdup_n_f16(half));
+        dst[i] = vgetq_lane_f32(fv, 0);
+    }
+
+#elif defined(__x86_64__)
+  #if defined(__AVX512F__)
+    // x86 AVX-512 bulk 16-wide
+    for (; i + 16 <= count; i += 16) {
```

Comment:
Is there a way/plan to add unit tests for the C++ implementation? That will give us more confidence about the correctness of the SIMD conversions. But I'm not sure how to integrate that into our existing unit testing framework, since not every machine running tests will have a supported processor.

Reply:
The only way I can think of is calling it via `JNICommons` in java and testing the conversions. I do have some edge case tests in a similar way in this PR.

Reply:
I think it's okay to test from the Java side. Could you do some manual/sanity testing later on with avx2,avx512, and neon instances? Something like a r5 instance will have avx512 and any graviton/apple m chip machine will have neon. M4 instance has avx2 and not avx512. I don't think our current CI tests across different processor platforms, but we want to be sure that the float conversions work on all supported platforms. Just want to be careful since I think this is the first PR (exciting!) that uses SIMD intrinsics directly in k-NN.